Sources/Instrumentation/Tracing/Span.swift

```diff
@@ -48,8 +48,65 @@ public protocol Span {
     /// - Parameter event: The `SpanEvent` to add to this `Span`.
     mutating func addEvent(_ event: SpanEvent)
 
-    /// The attributes describing this `Span`.
-    var attributes: SpanAttributes { get set }
+    /// Set a `String` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `String` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: String, forKey key: String)
+
+    /// Set a `[String]`value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `[String]` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: [String], forKey key: String)
+
+    /// Set a `Int` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `Int` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: Int, forKey key: String)
+
+    /// Set a `[Int]` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `[Int]` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: [Int], forKey key: String)
+
+    /// Set a `Double` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `Double` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: Double, forKey key: String)
+
+    /// Set a `[Double]` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `[Double]` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: [Double], forKey key: String)
+
+    /// Set a `Bool` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `Bool` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: Bool, forKey key: String)
+
+    /// Set a `[Bool]` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `[Bool]` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: [Bool], forKey key: String)
+
+    /// Set a `CustomStringConvertible` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `CustomStringConvertible` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: CustomStringConvertible, forKey key: String)
+
+    /// Set a `[CustomStringConvertible]` value for the given attribute key.
+    /// - Parameters:
+    ///   - value: The `[CustomStringConvertible]` value to set for the key.
+    ///   - key: The attribute key.
+    mutating func setAttribute(_ value: [CustomStringConvertible], forKey key: String)
```

Comment:
its more of a convenience method and does not define new type as such,

personally I'd add it to protocol extension (code for ilustration purposes, did not test it):

```swift
extension Span {
    public func setAnnotation(_ value: CustomStringConvertible, forKey key: String) {
       setAnnotation(String(describing: value), forKey: key)
    }
}
```

Reply:
BTW I think `ExpressibleByStringInterpolation` is very convinient for all "metadata",
that way you can use:

```swift
let result = true
let durationInSeconds: Float = 1.2 
setAnnotation("result: \(result), duration: \(durationInSeconds)s"), forKey: "key")
```

Reply:
last but not least, `@autoclosure` is sometimes handy, example:

```swift
    public func setAnnotation(_ value: @autoclosure () -> String, forKey key: String)
```

 (maybe let's leave it for now as an idea for further refinements/improvements)

Reply:
Yeah maybe it's worth doing all those as autoclosures, same as we did in the logging API.
It may be a bit weird to do that for a simple Int etc tbh but if we do we should do so consistently for all set... funcs.

It would save some allocs when not recording.

Reply:
> `setAnnotation("result: \(result), duration: \(durationInSeconds)s"), forKey: "key")`

that's just a String, no need to use expressible by string interpolation here.

Esp with an autoclosure there's nothing to gain there, we're already delaying rendering the string with the closure.

Reply:
That's a great point, thanks @pokryfka I'll add both `CustomStringConvertible` methods in an extension 👍

Reply:
I think @ktoso is right `ExpressibleByStringInterpolation` should be not needed if argument type is `String` itself,
just see (in unit test?) if `setAttribute("test \(1)", forKey key: "key")`
`String` uses different initializers for string literals and if it uses string interpolation both should work

Reply:
@pokryfka Not sure what you mean by that. Isn't it clear that when using `String` interpolation the `String` `setAttribute` method is being used?